include/sparrow/nullable.hpp

```diff
@@ -0,0 +1,589 @@
+
+        template <class To, class... Args>
+        concept assignable_from_one =
+            (std::assignable_from<std::add_lvalue_reference_t<To>, Args> && ...);
+
+        template <class T, class Arg>
+        concept assignable_from_refs =
+            assignable_from_one<T, Arg&, const Arg&, Arg&&, const Arg&&>;
+
+        template <class T, class U>
+        using conditional_ref_t = std::conditional_t<std::is_reference_v<T>, const std::decay_t<U>&, std::decay_t<U>&&>;
+
+        template <class T, class Targs, class U, class UArgs>
+        concept both_constructible_from_cond_ref =
+            std::constructible_from<T, conditional_ref_t<T, Targs>> and
+            std::constructible_from<U, conditional_ref_t<U, UArgs>>;
+
+        template <class To1, class From1, class To2, class From2>
+        concept both_convertible_from_cond_ref =
+            std::convertible_to<conditional_ref_t<To1, From1>, To1> and
+            std::convertible_to<conditional_ref_t<To2, From2>, To2>;
+
+        template <class To1, class From1, class To2, class From2>
+        concept both_assignable_from_cref =
+            std::assignable_from<std::add_lvalue_reference_t<To1>, mpl::add_const_lvalue_reference_t<From1>> and
+            std::assignable_from<std::add_lvalue_reference_t<To2>, mpl::add_const_lvalue_reference_t<From2>&>;
+
+        template <class To1, class From1, class To2, class From2>
+        concept both_assignable_from_cond_ref =
+            std::assignable_from<std::add_lvalue_reference_t<To1>, conditional_ref_t<To1, From1>> and
+            std::assignable_from<std::add_lvalue_reference_t<To2>, conditional_ref_t<To2, From2>>;
+
+        template <class T>
+        static constexpr bool is_nullable_v = mpl::is_type_instance_of_v<T, nullable>;
+    }
+
+    /**
+     * The nullable class.
+     */
+    template <class T, boolean_like B = bool>
+    class nullable
+    {
+    public:
+
+        using self_type = nullable<T, B>;
+        using value_traits = nullable_traits<T>;
+        using value_type = typename value_traits::value_type;
+        using reference = typename value_traits::reference;
+        using const_reference = typename value_traits::const_reference;
+        using rvalue_reference = typename value_traits::rvalue_reference;
+        using const_rvalue_reference = typename value_traits::const_rvalue_reference;
+        using flag_traits = nullable_traits<B>;
+        using flag_type = typename flag_traits::value_type;
+        using flag_reference = typename flag_traits::reference;
+        using flag_const_reference = typename flag_traits::const_reference;
+        using flag_rvalue_reference = typename flag_traits::rvalue_reference;
+        using flag_const_rvalue_reference = typename flag_traits::const_rvalue_reference;
+        
+        constexpr nullable() noexcept
+            : m_value()
+            , m_flag(false)
+        {
+        }
```

Comment:
We should only create the default constructor -  and make nullable default-constructible if `T` and `B` are default constructible.

If they are not, the std::variant will also not be. This may solve the issue with the follow up PR.

Reply:
Notice this does not fix the followup PR, but we still need to do it. Otherwise, `std::default_constructible<nullable<double&, bool>>` is `true`, even if trying to default construct it leads to a compilation error.